rfd/0138-postgres-backend.md

```diff
@@ -0,0 +1,132 @@
+
+As we intend to use logical decoding, we want to avoid transactions that are too big (in terms of rows changed); the only operations that affect more than one row are `DeleteRange` (which, in actual operation, is only ever used to delete a handful of rows at a time, for things like deleting a user's properties, password and MFA devices at once) and the deletion of expired items, which we'll have to do ourselves. Since logical decoding starts having performance issues when reaching the _thousands_ of rows changed at once, we only need to be careful when doing expired item deletion. Since we have no requirement that the deletion happens atomically (or in any specific order), we will take care to apply the operation across multiple transactions, limiting the amount of rows affected in each - such a mass expiration can happen, for instance, with the heartbeats of a large amount of nodes that have lost connection to the Teleport cluster.
+
+Each auth server running the backend will periodically attempt to delete expired items - at an interval tentatively set to 30 seconds - deleting small batches of expired items in different transactions until no more items can be deleted.
+
+To generate an event stream we'll connect to the database with a regular connection, create a temporary logical replication slot, then poll changes on it via the [`pg_logical_slot_get_changes`](https://www.postgresql.org/docs/current/functions-admin.html#FUNCTIONS-REPLICATION) function. As the builtin `pgoutput` logical decoding plugin cannot be used without errors via the SQL interface except in very recent versions of Postgres (11.20, 12.15, 13.11, 14.8 or 15.3, released in May 2023), we use the third-party but very popular [`wal2json`](https://github.com/eulerto/wal2json) plugin, which is available by default on the managed Postgres offerings in Azure, AWS and GCP, and is installable from the Debian/Ubuntu official packages or from the Postgres APT and YUM repos.
+
+The `pg_logical_slot_get_changes` function (with the appropriate `wal2json` options) will return one entry per change, which maps nicely to the backend event model; updates and inserts will be rendered as `OpPut` events, deletes as `OpDelete` events.
+
+### General architecture of the audit log
+
+The database schema for the audit log will be composed of just an `events` table with the appropriate indexes for searching:
+
+```pgsql
+CREATE TABLE events (
+    event_time timestamptz NOT NULL,
+    event_id uuid NOT NULL,
+    event_type text NOT NULL,
+    session_id uuid NOT NULL,
+    event_data json NOT NULL,
+    creation_time timestamptz NOT NULL DEFAULT now(),
+    CONSTRAINT events_pkey PRIMARY KEY (event_time, event_id)
+);
+CREATE INDEX events_creation_time_idx ON events (creation_time);
```

Comment:
Maybe look into using `brin` type of index (`CREATE INDEX events_creation_time_idx ON events USING BRIN (creation_time) WITH (pages_per_range=32);`? It might save a lot of index space and creation time is perfect type of data for this index type.

Reply:
Thank you for the advice, I'm switching the creation time index to BRIN.

The performance of event data cleanup doesn't matter much, so even the default `pages_per_range` (128) will probably work no matter what. Did you do some napkin math to pick 32 given the table layout?